gazebo_plugins/src/gazebo_ros_imu_sensor.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <gazebo_plugins/gazebo_ros_imu_sensor.hpp>
+#include <gazebo_ros/conversions.hpp>
+#include <gazebo_ros/utils.hpp>
+
+#include <iostream>
+#include <memory>
+#include <string>
+
+namespace gazebo_plugins
+{
+
+class GazeboRosImuSensorPrivate
+{
+public:
+  /// Node for ros communication
+  gazebo_ros::Node::SharedPtr ros_node_;
+  /// Publish for imu message
+  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr pub_;
+  /// IMU message modified each update
+  sensor_msgs::msg::Imu::SharedPtr msg_;
+  /// IMU sensor this plugin is attached to
+  gazebo::sensors::ImuSensorPtr sensor_;
+  /// Event triggered when sensor updates
+  gazebo::event::ConnectionPtr sensor_update_event_;
+
+  /// Publish latest imu data to ROS
+  void OnUpdate();
+};
+
+GazeboRosImuSensor::GazeboRosImuSensor()
+: impl_(std::make_unique<GazeboRosImuSensorPrivate>())
+{
+}
+
+GazeboRosImuSensor::~GazeboRosImuSensor()
+{
+}
+
+void GazeboRosImuSensor::Load(gazebo::sensors::SensorPtr _sensor, sdf::ElementPtr _sdf)
+{
+  impl_->ros_node_ = gazebo_ros::Node::Create("gazebo_ros_imu_sensor", _sdf);
+
+  impl_->sensor_ = std::dynamic_pointer_cast<gazebo::sensors::ImuSensor>(_sensor);
+  if (!impl_->sensor_) {
+    RCLCPP_ERROR(impl_->ros_node_->get_logger(), "Parent is not an imu sensor. Exiting.");
+    return;
+  }
+
+  impl_->pub_ = impl_->ros_node_->create_publisher<sensor_msgs::msg::Imu>("~/out");
+
+  // Create message to be reused
+  auto msg = std::make_shared<sensor_msgs::msg::Imu>();
+
+  // Get frame for message
+  msg->header.frame_id = gazebo_ros::SensorFrameID(*_sensor, *_sdf);
+
+  // Fill covariances
+  // TODO(anyone): covariance for IMU's orientation once this is added to gazebo
+  using SNT = gazebo::sensors::SensorNoiseType;
+  msg->angular_velocity_covariance[0] =
+    gazebo_ros::NoiseVariance(impl_->sensor_->Noise(SNT::IMU_ANGVEL_X_NOISE_RADIANS_PER_S));
```

Comment:
it's caused by https://bitbucket.org/osrf/gazebo/issues/2513/ in general, but even if the noise is specified in the SDF, the error still shows in tests because the noise isn't applied via the ServerFixture: https://github.com/ros-simulation/gazebo_ros_pkgs/issues/799